openmmtools/states.py

```diff
@@ -2012,23 +2089,69 @@ def _read_context_state(self, context_state, check_consistency):
+                                                  getVelocities=not ignore_velocities,
+                                                  getEnergy=not (ignore_velocities and ignore_positions),
                                                   enforcePeriodicBox=system.usesPeriodicBoundaryConditions())
         else:
+            if not ignore_collective_variables:
+                raise SamplerStateError("State objects must have ignore_collective_variables=True because they "
+                                        "don't track CV's and would be ambiguous between a System with no "
+                                        "collective variables.")
             openmm_state = context_state
 
-        positions = openmm_state.getPositions(asNumpy=True)
-        velocities = openmm_state.getVelocities(asNumpy=True)
-
         # We assign positions first, since the velocities
         # property will check its length for consistency.
-        self._set_positions(positions, from_context=True, check_consistency=check_consistency)
-        self._set_velocities(velocities, from_context=True)
-        self.box_vectors = openmm_state.getPeriodicBoxVectors(asNumpy=True)
         # Potential energy and kinetic energy must be updated
         # after positions and velocities or they'll be reset.
-        self.potential_energy = openmm_state.getPotentialEnergy()
-        self.kinetic_energy = openmm_state.getKineticEnergy()
+        if not ignore_positions:
+            positions = openmm_state.getPositions(asNumpy=True)
+            self._set_positions(positions, from_context=True, check_consistency=check_consistency)
+            self._potential_energy = openmm_state.getPotentialEnergy()
+        if not ignore_velocities:
+            velocities = openmm_state.getVelocities(asNumpy=True)
+            self._set_velocities(velocities, from_context=True)
+            self._kinetic_energy = openmm_state.getKineticEnergy()
+        self.box_vectors = openmm_state.getPeriodicBoxVectors(asNumpy=True)
+        # This has its own function for handling ignore_cv's because States have different behavior
+        self._read_collective_variables(context_state, ignore_collective_variables)
+
+    def _read_collective_variables(self, context_state, ignore_collective_variables):
+        """
+        Update the collective variables from the context object
+
+        Parameters
+        ----------
+        context_state : simtk.openmm.Context or simtk.openmm.State
+            The object to read. This only works with Context's for now,
+            but in the future, this may support OpenMM State objects as well.
+        ignore_collective_variables : bool
+            Choose to just ignore this value
+        """
+        if ignore_collective_variables:
+            if isinstance(context_state, openmm.State):
+                # Invalidate if State is passed in
+                self._collective_variables = None
```

Comment:
I'm not sure we want to do this. I'm worried about the fact that these two snippets:
```python
sampler_state.update_from_context(context_state, ignore_collective_variables=True)
sampler_state.update_from_context(context, ignore_positions=True, ignore_velocities=True)
```
and
```python
sampler_state.update_from_context(context, ignore_positions=True, ignore_velocities=True)
sampler_state.update_from_context(context_state, ignore_collective_variables=True)
```
will result in two different states. I think this mistake would be hard to find and it's easy to miss this behavior in the docs. I think it's safer to have always `ignore_collective_variables` leave the CV as it is because if the user is using the `ignore_X` attributes it means that he's updating everything by steps.

Hmm. I understand now that you were probably worried about the fact that the user won't be able to invalidate a particular attribute during the manual update. If this is the case, I'd rather have the properties setters allow to set (exclusively) the `None` value.

Reply:
Because we support both `State` and `Context` objects, which now cause a loss of information, we have to do something to handle it. I don't want the user to by default be able to put the `SamplerState` into something inconsistent.

>  I understand now that you were probably worried about the fact that the user won't be able to invalidate a particular attribute during the manual update.

When would this be?

> I'd rather have the properties setters allow to set (exclusively) the None value.

I can change that, but lets figure out all the use cases first.

Reply:
> I don't want the user to by default be able to put the SamplerState into something inconsistent.

This is already possible if the user does
```python
sampler_state.update_from_context(context, ignore_collective_variables)
```
and `context` is an actual `Context` object. We must assume that users know what they're doing when they are requesting only partial information to optimize the communication with the GPU, but I'm worried an inconsistent behavior of the input arguments would trick also more advanced users (me included).